File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        helper(ans, "", 0, 0, n);
        return ans;
    }

    public void helper(List<String> ans, String str1, int open, int close, int n){
        if(open == n && close == n){
            ans.add(str1);
        }
        if(open<n){
            helper(ans, str1+"(", open+1, close, n);
        }
        if(close<open){
            helper(ans, str1+")", open, close+1, n);
        }
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        self.helper(ans, "", 0, 0, n)
        return ans
    def helper(self,ans: List[str], str1:str, open:int, close:int, n:int):
        if open == n and close == n:
            ans.append(str1)
            return
        if open < n:
            self.helper(ans, str1+"(", open+1, close, n)
        if close < open:
            self.helper(ans, str1 + ")", open, close+1, n)
        